src/artm/core/master_component.cc

```diff
@@ -227,9 +357,335 @@ void MasterComponent::AttachModel(const AttachModelArgs& args, int address_lengt
   instance_->merger()->SetPhiMatrix(model_name, attached);
 }
 
+
 void MasterComponent::InitializeModel(const InitializeModelArgs& args) {
   LOG(INFO) << "MasterComponent::InitializeModel() with " << Helpers::Describe(args);
+
+  // temp code to be removed with ModelConfig and old dictionaries
   instance_->merger()->InitializeModel(args);
+
+  // ToDo: (MelLain) implement the initialization of new-style models with new-style dicts
+  // instance_->merger()->SetPhiMatrix(model_name, attached);
+}
+
+void MasterComponent::FilterDictionary(const FilterDictionaryArgs& args) {
+  LOG(INFO) << "MasterComponent::FilterDictionaryArgs() with " << Helpers::Describe(args);
+
+  auto src_dictionary_ptr = instance_->dictionary_impl(args.dictionary_name());
+  if (src_dictionary_ptr == nullptr) {
+    LOG(ERROR) << "MasterComponent::FilterDictionaryArgs(): filter was requested for non-exists dictionary '"
+      << args.dictionary_name() << "', operation was aborted";
+  }
+
+  std::string dictionary_target_name = args.has_dictionary_target_name() ? args.dictionary_target_name() :
+      "dictionary_" + boost::lexical_cast<std::string>(boost::uuids::random_generator()());
+  LOG(INFO) << "The name of filtered dictionary is '" << dictionary_target_name << "'";
+
+  auto dictionary_data = std::make_shared<artm::DictionaryData>();
+  dictionary_data->set_name(dictionary_target_name);
+
+  auto& src_entries = src_dictionary_ptr->entries();
+  auto& dictionary_token_index = src_dictionary_ptr->token_index();
+  std::unordered_map<int, int> old_index_new_index;
+
+  int accepted_tokens_count = 0;
+  for (auto& entry : src_entries) {
+    if (args.has_class_id() && entry.token().class_id != args.class_id()) continue;
+
+    if (args.has_min_df() && entry.token_df() < args.min_df()) continue;
+    if (args.has_max_df() && entry.token_df() >= args.max_df()) continue;
+
+    if (args.has_min_tf() && entry.token_tf() < args.min_tf()) continue;
+    if (args.has_max_tf() && entry.token_tf() >= args.max_tf()) continue;
+
+    if (args.has_min_value() && entry.token_value() < args.min_value()) continue;
+    if (args.has_max_value() && entry.token_value() >= args.max_value()) continue;
+
+    // all filters were passed, add token to the new dictionary
+    Token token = entry.token();
+    accepted_tokens_count += 1;
+    dictionary_data->add_token(token.keyword);
+    dictionary_data->add_class_id(token.class_id);
+    dictionary_data->add_token_df(entry.token_df());
+    dictionary_data->add_token_tf(entry.token_tf());
+    dictionary_data->add_token_value(entry.token_value());
+
+    old_index_new_index.insert(std::pair<int, int>(dictionary_token_index.find(token)->second,
+                                                   accepted_tokens_count - 1));
+  }
+
+  auto cooc_dictionary_data = std::make_shared<artm::DictionaryData>();
```

Comment:
Wow.. This approach require you to implement filtering logic twice. I meant something simpler.. Just create a constructor of `::artm::core::Dictionary` class that accepts `::artm::DictionaryData` as a parameter. Then, use this constructor in `CreateOrReconfigureDictionaryImpl` to automatically create an old dictionary (every time the method is called).